File: benchmark/example_solver.py
#!/usr/bin/env python3
"""One-file submission template; replace solve() with your algorithm.

This example handles only identity inputs. It demonstrates the interface,
not a competitive realization algorithm. The benchmark owns all transport.
"""

from __future__ import annotations


def solve(c: list[float], g: list[float], t: list[float]) -> list[list[float]] | None:
    """Return a real SO(4) frame, or None when no frame was found."""
    if all(value == 0.0 for point in (c, g, t) for value in point):
        return [[float(i == j) for j in range(4)] for i in range(4)]
    return None

File: benchmark/example_solver.rs
//! Minimal submission: solve identity inputs and decline everything else.
//!
//! Replace this function with your algorithm. Available dependencies are
//! std, nalgebra 0.35, and serde_json 1. Keep stdout free for the benchmark.

/// Return a real SO(4) frame, or None when no frame was found.
pub fn solve(c: [f64; 3], g: [f64; 3], t: [f64; 3]) -> Option<[[f64; 4]; 4]> {
    if [c, g, t].iter().flatten().all(|value| *value == 0.0) {
        Some(std::array::from_fn(|i| {
            std::array::from_fn(|j| if i == j { 1.0 } else { 0.0 })
        }))
    } else {
        None
    }
}
